client/src/tpu_connection.rs

```diff
@@ -10,12 +10,15 @@ pub trait TpuConnection {
 
     fn tpu_addr(&self) -> &SocketAddr;
 
-    fn send_transaction(&self, tx: &Transaction) -> TransportResult<()> {
-        let data = bincode::serialize(tx).expect("serialize Transaction in send_transaction");
-        self.send_wire_transaction(&data)
-    }
-
-    fn send_wire_transaction(&self, data: &[u8]) -> TransportResult<()>;
+    fn send_wire_transaction(&self, wire_transaction: &[u8]) -> TransportResult<()>;
 
-    fn send_batch(&self, transactions: &[Transaction]) -> TransportResult<()>;
+    fn send_wire_transaction_batch(
```

Comment:
I think this change can be much smaller. For example, I think we can simply implement a new send_wire_transaction_batch mirroring the send_wire_transaction function already present. I also disagree with the way send_wire_transaction_batch here is implemented. If you look at the original send_batch function, it was not implemented by simply looping over all the transactions and calling send_transaction on them. It used a specialized implementation that used multiple concurrent streams and stuff. This was found to be necessary, as when sending large numbers of transactions, not taking advantage of multiple concurrent streams was found to be unacceptably slow. So I would prefer a much smaller change that simply implemented a new function like

     fn send_wire_transaction_batch(&self, data: &[Vec<u8>]) -> TransportResult<()> {
        let _guard = self.client.runtime.enter();
        let send_batch = self.client.send_batch(data);
        self.client.runtime.block_on(send_batch)?;
        Ok(())
    }

Reply:
Maybe we should templatize the wire interface(s) on `AsRef<[u8]>` and let any callers prepare their data however is appropriate

Reply:
> I think this change can be much smaller

Yes, it can be smaller but can you explain why you want the change to be smaller? I personally feel that keeping the `TpuConnection` interface smaller is a better goal than minimizing changes to this interface since it's still quite new.

> I also disagree with the way send_wire_transaction_batch here is implemented. If you look at the original send_batch function, it was not implemented by simply looping over all the transactions and calling send_transaction on them. It used a specialized implementation that used multiple concurrent streams and stuff.

I think you're misunderstanding this change. Previously, `UdpClient` was using a simple iterator to send transactions for a batch. I moved that implementation to the default implementation of the `TpuConnection` trait and the `QuicClient` overrides that default implementation to use concurrent streams.

Reply:
> Maybe we should templatize the wire interface(s) on AsRef<[u8]> and let any callers prepare their data however is appropriate

I'd prefer to just keep this interface simple and just pass through a `&[u8]` since that's what the `UdpSocket` and `QuicClient` are already using. But I don't feel that strongly, could do what you propose or have a new `IntoWireTransaction` trait that we use here.

Reply:
Got it, makes sense. One thing i'd still like to keep around though is a send_transaction_batch does maybe just serializes the transaction and calls send_wire_transaction_batch for convenience.